### PR TITLE
feat: add interactive NX GUI bridge as a C# add-in

### PR DESCRIPTION
**Pull request by mikbalarikan**
## Summary

`docs/architecture.md` names a fallback for NX builds that cannot provide a non-blocking GUI scheduler: a minimal C# NX-side bridge. This PR implements it. `nx_gui_bridge/` adds an NX Open .NET add-in that serves bridge protocol v1 from an interactive NX session. NX stays usable, and the model updates on screen while MCP tools run.

These stay unchanged:

- the sidecar and the 16 certified tools;
- the descriptor, the JSON-RPC envelope and the error codes.

The only Python change is the optional `NX_MCP_STATE_DIR` variable. Most of the diff is one C# file of 2,348 lines.

## Why

- The bundled runner pumps requests in a blocking journal, so the NX UI freezes. It works in batch only.
- A Python UI-thread scheduler does not work on NX2206:
  - NX runs Python startup scripts in a sub-interpreter (ID 1), and its UI thread holds the GIL while idle.
  - A ctypes Win32 timer callback deadlocked that thread on its first tick.
  - Every responsiveness probe over 20 s reported the window as hung.

## Changes

**1. `feat: allow NX_MCP_STATE_DIR to relocate the bridge descriptor`** (`src/nx_mcp/bridge.py` and tests)

- When `NX_MCP_STATE_DIR` is set, `default_descriptor_path()` returns `$NX_MCP_STATE_DIR/bridge.json`. The default location is unchanged.
- An MSIX-packaged MCP client, and every process it starts, sees a private copy of `%LOCALAPPDATA%`. Its sidecar therefore never finds the `bridge.json` written by an NX started from Explorer. On the test machine, the Claude desktop app was such a client.
- The tests pin the variable, so a developer's environment cannot point them at a live bridge.

**2. `feat: add interactive NX GUI bridge as a C# add-in`** (`nx_gui_bridge/` and docs)

- A background .NET thread owns the loopback listener with the same limits as `nx_mcp.bridge`:
  - `127.0.0.1` on a random port;
  - a 256-bit token;
  - 1 MiB per message and a 5 s receive deadline.

  Requests are handled one at a time.
- Each call runs on NX's UI thread through a hidden WinForms control. It starts only after `UF_UI_lock_ug_access(UF_UI_FROM_CUSTOM)` returns `UF_UI_LOCK_SET`.
  - While a dialog or command owns NX, the call retries every 250 ms.
  - After 60 s it fails with `NX_MAIN_THREAD_UNAVAILABLE` (`not_started`, retryable).
- The executor is a C# port of `NXOpenExecutor` for the 16 certified commands. It keeps:
  - undo marks with rollback;
  - checks for stale and wrong-kind IDs;
  - workspace confinement on both sides;
  - the STEP export settings from #6.

  Legacy and experimental commands are not available.
- Configuration comes from `gui-bridge.json` next to the DLL. If that file is missing, the bridge reads `%LOCALAPPDATA%\nx-mcp\gui-bridge.json`, and environment variables override both.
  - `workspace` is required.
  - `state_dir` and `stop_file` are optional.
- Two ways to load it:
  - Run File > Execute > NX Open (Ctrl+U). Running it again stops the bridge.
  - Put it in the `startup` folder of a custom directory.
- The bridge logs each method, its outcome and its duration. The token is never logged.
- `build.ps1` compiles with the in-box .NET Framework `csc.exe` (C# 5), so no SDK is needed. NX loads an unsigned add-in only with a `dotnet_author` license; otherwise, sign the DLL with `SignDotNet.exe`.
- Docs:
  - The new `docs/gui-bridge.md` covers the build, configuration, limits and the validation record.
  - The README capability row is updated, and a short README section is added.
  - `docs/architecture.md` points to the add-in.

## Validation

Local checks (Windows 11, Python 3.12.13):

- `pytest -q -p no:cacheprovider -m "not real_nx"`: 299 passed, 14 skipped. Branch coverage is 82.95%.
- `ruff check`, `ruff format --check` and `mypy src/nx_mcp` are clean.
- CI does not build the C# add-in.

Real NX was **NX2206 build 2206.9101 only**, on 2026-09-18. The add-in was built with the in-box .NET Framework 4.8 `csc.exe`, without an SDK. The runs used a branch that combined this PR and #6. Its tree is identical to the result of merging both PRs.

| Check | Result |
| --- | --- |
| Load from a custom directory's `startup` folder | NX called `Startup()` on the UI thread. |
| `python -m nx_mcp.real_smoke --iterations 1` | Passed in 29 s. |
| `pytest -m real_nx tests/test_real_nx.py` (20 iterations plus negative cases) | Passed in 297 s. Each of the 20 STEP files had one solid, 6 faces and a 20 × 10 × 12.5 mm extent. |
| Protocol parity | For 21 malformed and edge-case requests, the codes and `execution_state` matched the Python server. |
| Stop file | The descriptor was removed. Later calls returned `NX_BRIDGE_UNAVAILABLE` (`not_started`). |
| NX started from Explorer, sidecar inside an MSIX-packaged client, shared `state_dir` | The sidecar found the bridge at once. A 1-iteration smoke passed in 19 s. |
| A live MCP session in that interactive NX | An L-bracket (6 sketch lines, 30 mm extrude) was modeled, saved and exported to STEP in 46 s. The STEP had 1 solid, 8 faces and a 60 × 40 × 30 mm extent. |

UI responsiveness was probed every 250 ms during the whole acceptance run: 1,491 probes over 570 s.

- Windows never reported the window as hung.
- The median probe latency was 0.26 ms.
- NX was busy only while a single command ran. The longest busy stretch was 4.5 s, and 28 probes timed out at 2 s, all during commands.

## Not yet exercised

- deferral while a user dialog is open;
- the Ctrl+U start/stop toggle;
- running alongside other in-process NX plugins;
- NX 2506.

## Notes for reviewers

- `.github/workflows/real-nx.yml` still polls `%LOCALAPPDATA%\nx-mcp\bridge.json`. This only matters if a runner sets `NX_MCP_STATE_DIR`.
- This PR does not depend on #6, because the add-in has its own port of the STEP settings. The solid check in the real-NX runs above comes from #6.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

- **New Features**
  - Added an interactive NX GUI bridge for live modeling in an open NX session.
  - Supports certified actions such as creating, opening, saving, exporting, sketching, extruding, undoing, and viewing workspace information.
  - Added configurable bridge state storage through `NX_MCP_STATE_DIR`; the path must be absolute.

- **Documentation**
  - Documented NX2206 compatibility, setup requirements, validation results, behavior, and limitations.
  - Clarified graceful shutdown behavior so in-progress responses complete before disconnecting.

- **Tests**
  - Updated state-directory tests for explicit paths, invalid relative paths, and fallback behavior.

### CODE REVIEW

**Comment by DreamEnding** [comment]
Thanks for putting this together. The amount of real-NX validation here is especially useful, including the interactive NX2206 run, protocol checks, and UI responsiveness measurements. The C# bridge also addresses an important limitation of the current batch-oriented Python bridge.

I reviewed the implementation in detail and found a few areas that I would like to see addressed before merging.

First, the STEP export code has the same stale-output edge case as PR #6. If the destination STEP already exists and the translator fails to write new output, the existing non-empty file can still make the operation appear successful. Please update the C# implementation together with the Python implementation so both bridges have the same failure semantics.

Second, `ConfigCandidates()` currently uses `AppDomain.CurrentDomain.BaseDirectory` as the preferred location for `gui-bridge.json`. That directory is the application domain base directory and is not necessarily the directory containing `NxMcpGuiBridge.dll` when the add-in is loaded by NX. Since the documentation says that a config placed next to the DLL is supported, I think the implementation should derive that location directly from `typeof(BridgeHost).Assembly.Location`.

Third, I am concerned about the deferred-stop path while an NX call is executing. `Stop()` currently reposts itself through `BeginInvoke()` whenever `dispatcher.Executing` is true. If NX is inside a nested UI message loop, the reposted callback may run before the current NXOpen call returns and repeatedly repost itself. A pending-stop flag/state would be safer: record the stop request while execution is active and perform the actual shutdown once the current call has completed.

There are also two follow-up improvements that I strongly recommend, even if they do not necessarily need to block the same commit:

* `RefuseLiveDescriptor()` currently treats any process listening on the stored port as an existing NX MCP bridge. It would be more robust to verify the descriptor by performing an authenticated protocol request using the stored token, so a stale descriptor whose port has been reused by another process does not prevent startup.
* The C# add-in is now a substantial part of the implementation, but the hosted CI does not compile it. Please consider adding at least a build/compile gate, and ideally unit coverage for the protocol/configuration/dispatcher logic that does not require a live NX installation.

The overall direction looks reasonable, and the NX2206 validation gives useful evidence that the approach works in practice. I would be happy to take another look after these issues are addressed. Thanks again for the substantial contribution.

**Comment by mikbalarikan** [comment]
Thanks for the detailed review. All three blocking points are addressed, and the first follow-up as well.

**1. STEP export (0164af1).** `ExportStep` now translates into a temporary directory beside the destination and moves the result into place only once that file exists and is non-empty, so both bridges fail the same way. The Python side is c0a52f9 in #6. Two details that shaped it:

- The file keeps its requested name inside the temporary directory, because NX records the output name in the STEP header. A temporary *file* name put `part.nx-mcp-<random>.stp` into `FILE_NAME`; with a directory the output is byte for byte what it was before.
- NX writes its translator log next to the output, so the log is moved to the destination's name. Cleanup removes the temporary directory and never masks the failure that caused it.

**2. Config location (0164af1).** `ConfigCandidates()` derives the add-in folder from `typeof(BridgeHost).Assembly.Location` and falls back to the application domain base directory only when that location is empty.

**3. Deferred stop (0164af1).** `Stop()` no longer reposts itself. A stop requested while a call runs records the reason, calls `dispatcher.Stop()` so no new work is accepted, and returns. The dispatcher invokes a completion hook after the running call returns, and that posts the shutdown exactly once. The flag is cleared on start, so a stop that never completed cannot stop the next bridge.

One thing worth reporting: I could not exercise that path automatically. The stop-file watcher polls between requests, so while a call runs the accept loop is inside `Respond` and the watcher cannot fire: it only ever stops an idle bridge. The reachable trigger is Ctrl+U, or another in-process caller, during a call, which needs interactive input. `docs/gui-bridge.md` now says so instead of claiming coverage.

**4. Descriptor verification (127d4ce).** The probe sends an authenticated request with the stored token and refuses to start only when the answer is a bridge response, an authentication failure included. Anything else is logged as a stale descriptor and ignored.

## Re-validation on NX2206 (build 2206.9101, today)

The add-in was rebuilt with the in-box `csc.exe` (no warnings at `/warn:4`) and loaded from the `startup` folder of a test NX:

| Check | Result |
| --- | --- |
| `pytest -m real_nx tests/test_real_nx.py`, 20 iterations plus the negative cases | Passed in 435 s. All 20 STEP files have one solid, `FILE_NAME` stayed `run-01.stp`, and the sizes match the runs before these changes. |
| Three exports to the same path: a fresh part, over the existing file, and after a save | Each replaced the destination with a one-solid STEP and left no temporary directory behind. |
| Descriptor whose port an unrelated service holds | Logged as stale and ignored; the bridge started. The old probe refused here. |
| Descriptor served by a live bridge | Start refused, naming that bridge's pid and port, and left the descriptor untouched. |
| Stop file | Descriptor removed. |

Python checks are unchanged: 299 passed, 14 skipped, branch coverage 83.02%; ruff, ruff format and mypy clean.

## On the CI gate

A hosted runner cannot compile the add-in: `csc` needs `NXOpen.dll`, `NXOpen.UF.dll`, `NXOpenUI.dll` and `NXOpen.Utilities.dll` from an NX installation. Three ways to get a gate, and I am happy to implement whichever you prefer:

1. Add the `build.ps1` compile step to `.github/workflows/real-nx.yml`, which already runs on the self-hosted Windows runner with NX. Cheapest, and it gates every real-NX run.
2. Add NX-free coverage that runs on hosted CI: a test that parses `nx_gui_bridge/NxMcpGuiBridge.cs` and asserts the protocol constants (version, the 1 MiB message limit, the 5 s receive deadline) and the certified command list match `nx_mcp.bridge` and `nx_mcp.nx_bridge`. It catches drift between the two bridges on every PR without NX.
3. Split the NX-free logic (JSON, framing, workspace, config) into its own file and compile plus unit test that part on hosted CI. Real unit coverage, but it adds a .NET SDK and a C# test framework to the repository.

My suggestion is 1 and 2 now, and 3 only if you want C# unit tests in the tree. Say which you would like and I will add it here or in a follow-up PR.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: nx_gui_bridge/NxMcpGuiBridge.cs [hunk 1]
```diff
@@ -0,0 +1,2497 @@
...
+                    request["jsonrpc"] = "2.0";
+                    request["id"] = "nx-mcp-descriptor-probe";
+                    request["protocol_version"] = Protocol.Version;
+                    request["token"] = secret == null ? "" : secret;
+                    request["method"] = "nx_status";
+                    request["params"] = new Dictionary<string, object>();
+                    byte[] payload = Protocol.Utf8.GetBytes(Json.Serialize(request) + "\n");
+                    NetworkStream stream = probe.GetStream();
+                    stream.WriteTimeout = 1000;
+                    stream.ReadTimeout = 2000;
```
[the remark this answers is not in the document]
Comment (mikbalarikan) [reply nx_gui_bridge/NxMcpGuiBridge.cs]: Fixed in be2e851. The probe now sends a deliberately invalid token, which a bridge rejects with `NX_AUTH_FAILED` on its socket thread, so a busy bridge still answers; startup continues only on proof that the port is foreign (nothing listening, a connection that closes, or an answer that is not a bridge response), and silence leaves the descriptor alone. Verified on NX2206 against an idle bridge, a bridge held inside a 300 s call, and an unrelated service on the port.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: README.md [hunk 2]
```diff
@@ -56,7 +56,7 @@ there is no blanket certification of every tool or failure branch.
 | Authentication, workspace boundaries, stale/wrong-kind IDs | Enforced by default | Boundary and fake-NX tests | Current real-NX negative cases await rerun |
 | Uncertain execution and forced rollback failure | Enforced recovery rules | Local fault injection | No real-NX fault-injection acceptance claimed |
 | Bridge/process restart | Existing lifecycle | Automated bridge tests | Current independent-process restart test awaits rerun |
-| Interactive GUI scheduling | Not available in bundled runner | Batch runner only | Non-blocking UI scheduler or minimal C# bridge still required |
+| Interactive GUI scheduling | C# add-in in `nx_gui_bridge/`, loaded on demand | Protocol checks against the live add-in | NX2206 acceptance only (2026-09-18); see [GUI bridge](docs/gui-bridge.md) |
```
[the remark this answers is not in the document]
Comment (mikbalarikan) [reply README.md]: Fixed in be2e851: the capability row now reads "NX2206 acceptance only (2026-09-18, re-validated 2026-09-20)".

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: src/nx_mcp/bridge.py [hunk 3]
```diff
@@ -86,6 +86,11 @@ def write(self, path: str | Path) -> None:
 
 
 def default_descriptor_path() -> Path:
+    # MSIX-packaged MCP clients see a private copy of AppData, so the NX bridge
+    # and the sidecar may need an explicit shared directory outside it.
+    state_directory = os.environ.get("NX_MCP_STATE_DIR")
+    if state_directory:
+        return Path(state_directory) / "bridge.json"
```
[the remark this answers is not in the document]
Comment (mikbalarikan) [reply src/nx_mcp/bridge.py:101]: Fixed in be2e851: both sides now reject a relative `NX_MCP_STATE_DIR` / `state_dir` and keep treating an empty value as unset, with tests for the absolute, relative and empty cases in `tests/test_bridge.py`.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

**Comment by mikbalarikan** [comment]
Correction to the descriptor probe I described above. CodeRabbit pointed out that probing with the descriptor's real token makes a live bridge pass `nx_status` to NX's UI thread, so a bridge that is busy cannot answer inside the read deadline and startup would overwrite a live descriptor.

be2e851 probes with a deliberately invalid token instead, which a bridge rejects on its socket thread, and lets startup continue only on proof that the port belongs to something else: nothing listening, a connection that closes, or an answer that is not a bridge response. Silence is not proof.

Verified on NX2206 against three cases: an unrelated service holding the port (the bridge starts), an idle bridge (refused), and a bridge held inside a 300 s call (refused, descriptor untouched).

The same commit also requires `NX_MCP_STATE_DIR` and `state_dir` to be absolute on both sides, since a relative value resolved against each process's own working directory, and updates the README validation date.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

**Comment by DreamEnding** [comment]
Thanks for the updates and the additional validation. I re-reviewed the current `be2e851` head.

The earlier issues around stale STEP output, config resolution, descriptor ownership, busy-bridge probing, and absolute `state_dir` handling now look addressed. The current CI is also green, and the NX2206 re-validation is useful.

I still see two blocking issues in the C# bridge that I think should be fixed before merge:

1. **Deferred stop can close the active connection before the current response is sent.**

   After an NX call completes, `pending.Done.Set()` wakes the socket thread, but the completion hook can also queue `Stop()` immediately. `Stop()` then calls `server.Stop()`, which shuts down the current active socket.

   At that point, the socket thread may not yet have encoded and sent the RPC response. This creates a race where the NX mutation succeeds, but the client loses the success response and sees an uncertain execution instead.

   For a deferred stop, please reject new work immediately but allow the current request to finish sending its response before shutting down the active connection and completing bridge teardown.

2. **The response path still assumes one `Socket.Send()` sends the entire buffer.**

   The current code calls:

   ```csharp
   client.Send(encoded);
   ```

   but ignores the returned byte count. `Socket.Send()` may send fewer bytes than requested, so a larger response can be truncated.

   Please use send-all semantics, e.g. loop until all bytes have been transmitted or the connection fails.

These two issues both affect the same correctness property: a command may already have executed in NX while the sidecar fails to receive a complete response.

Once these two response/lifecycle issues are addressed, I don't currently see another blocking issue from my review.

Thanks again for the careful follow-up and the repeated NX2206 validation.

**Comment by mikbalarikan** [comment]
Thanks for the re-review, and for merging #6. Both issues are fixed in 7f73510.

**1. Deferred stop no longer cuts the response.** You were right that the hook and the socket thread raced. `Stop()` now only refuses new work when a call is running, and the completion hook hands teardown to `BridgeServer.StopAfterResponse()`, which:

- sets `stopping` so no new connection is taken, and closes the listener;
- leaves the active socket alone and joins the serve thread, so the in-flight request finishes encoding and sending;
- runs on a short-lived background thread, because the UI thread must not block, and posts the last step back to the UI thread, because the marshaling control has to be disposed by its owner.

A `stopInProgress` guard keeps a second stop request, for example from Ctrl+U, out of the teardown, and `FinishStop` is idempotent.

**2. Responses are sent in full.** The response path now loops until every byte of the encoded reply has gone out, with the existing 5 s `SendTimeout` still in place, which matches the `sendall` the Python server already does.

## Measured on NX2206 (build 2206.9101)

A stop that arrives during a call cannot come from the stop-file watcher: it polls between requests, so it only ever stops an idle bridge. To reach the path, I used a test build that calls `BridgeHost.Stop` from inside `nx_export_step` (not in the PR), and a second one that also waits 300 ms before sending the response, to widen exactly the window you described:

| Build | Result |
| --- | --- |
| Before this fix, widened window | The client got `NX_PROTOCOL_ERROR` with `execution_state: unknown`, while NX had already written the STEP with one solid. Your race, reproduced. |
| Before this fix, no artificial delay | The response happened to win by about 50 ms, which is why the acceptance runs never caught it. |
| With this fix, widened window | `Exported STEP: stopmid.stp` reached the client in full, then `stopped`. |
| With this fix, no artificial delay | Same, response first and teardown after. |

The shipped build also passed `python -m nx_mcp.real_smoke --iterations 2` followed by a stop through the stop file, with the descriptor removed. `docs/gui-bridge.md` records all of this, and the Ctrl+U route into the same path is still listed as not exercised.

Python checks are unchanged: 301 passed, 14 skipped, branch coverage 82.98%; ruff, ruff format and mypy clean.

A partial `Socket.Send` is hard to force on loopback, so that change rests on review rather than a test; if you would like a stress case for it, I can add one to the acceptance script.

The CI gate question from my earlier comment is still open whenever you want to pick an option.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: nx_gui_bridge/NxMcpGuiBridge.cs [hunk 4]
```diff
@@ -0,0 +1,2620 @@
...
+            catch (Exception)
+            {
+                return;
+            }
+            object port;
+            if (existing == null || !existing.TryGetValue("port", out port) || !(port is long))
+            {
+                return;
+            }
+            if (!IsForeignPort((int)(long)port))
```
[the remark this answers is not in the document]
Comment (mikbalarikan) [reply nx_gui_bridge/NxMcpGuiBridge.cs]: Fixed in 61f457f: a descriptor port outside 1..65535 is now logged as stale and ignored before the cast, matching `BridgeDescriptor.read` on the Python side. Verified on NX2206 with a descriptor carrying port 99999: the add-in logged "port 99999 is out of range" and started instead of reporting a live bridge.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

**Comment by mikbalarikan** [comment]
Both findings from the latest CodeRabbit review are fixed in 61f457f. The second one had no inline thread, so here it is.

**`state_dir` and `bridge.json` permissions.** The point is fair: the descriptor carries the session token, and `state_dir` can now be pointed outside the user profile, so the add-in no longer relies on inherited rules.

- A missing state directory is created with a single rule for the current account, with inheritance turned off.
- `bridge.json` is created for that account alone before any content is written, and, because replacing a file keeps the destination's rules, the rule is set on the final path as well.
- The log inherits the directory rule and never contains the token.
- A state directory that already exists keeps the rules it has, so `docs/gui-bridge.md` and the README now say plainly that `state_dir` must not be a folder other accounts can read. The Python bridge already chmods the descriptor to 0600 where the platform supports it.

Measured on NX2206 with a state directory the add-in had to create:

```
state-acl        MERY\marya:(OI)(CI)(F)
bridge.json      MERY\marya:(F)
gui-bridge.log   MERY\marya:(I)(F)
```

No other account or group appears, and a 1-iteration smoke through that bridge still passed.

**Descriptor port range.** An out-of-range port is treated as a stale descriptor before the cast, as `BridgeDescriptor.read` already does in Python. With a planted descriptor on port 99999 the add-in logged `port 99999 is out of range` and started.

Local checks: 301 passed, 14 skipped, branch coverage 83.05%; ruff, ruff format and mypy clean.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

**Comment by DreamEnding** [comment]
Thanks for the work on this PR. I’ve merged it.

The overall implementation is solid, and I especially appreciate the amount of real NX2206 validation and the follow-up fixes around bridge probing, shutdown behavior, descriptor permissions, and port validation.

There is one concurrency issue I’d still like to address in a follow-up PR: the descriptor startup path currently has a cross-process TOCTOU window between checking the existing `bridge.json` and publishing the new descriptor. Two NX processes starting the GUI bridge at nearly the same time could both pass the initial check and race to claim the descriptor. The shared `bridge.json.tmp` path should also be made unique per writer.

This does not block the merge anymore, but I’d like us to fix it next by protecting the full:

`RefuseLiveDescriptor() -> listener.Start() -> WriteDescriptor()`

sequence with an inter-process startup lock, and by using a unique temporary descriptor file.

Thanks again for the contribution and the detailed validation work.
